moped-editor/src/views/projects/projectView/ProjectComponentEdit.js

```diff
@@ -0,0 +1,777 @@
+          })
+        )
+      : [];
+
+  /**
+   * This is a constant list containing the names of available types
+   * @type {String[]}
+   */
+  const availableTypes = data
+    ? [
+        ...new Set(
+          data.moped_components.map(
+            moped_component => moped_component.component_name
+          )
+        ),
+      ].sort()
+    : [];
+
+  /**
+   * Generates a list of available subtypes for a fiven type name
+   * @param {String} type - The type name
+   * @return {String[]} - A string array with the available subtypes
+   */
+  const getAvailableSubtypes = type =>
+    Object.entries(initialTypeCounts[type]?.subtypes ?? {})
+      .map(([_, component]) => (component?.component_subtype ?? "").trim())
+      .filter(item => item.length > 0);
+
+  /**
+   * Returns any available subcomponent objects for a specific type
+   * @param {String} type - Lower case of the name of the type
+   * @return {Object[]}
+   */
+  const getAvailableSubcomponents = () =>
+    initialTypeCounts[selectedComponentType].subcomponents;
+
+  /**
+   * Returns the current project componetn id
+   * @return {number|null}
+   */
+  const getProjectComponentId = () =>
+    data?.moped_proj_components[0]?.project_component_id ?? null;
+
+  /**
+   * Handles the delete button click
+   */
+  const handleDeleteDialogClickOpen = () => {
+    setDeleteDialogOpen(true);
+  };
+
+  /**
+   * Handles the closing of the dialog
+   */
+  const handleDeleteDialogClickClose = () => {
+    setDeleteDialogOpen(false);
+  };
+
+  /**
+   * On select, it changes the available items for that specific type
+   * @param {Object} e - The Object Type
+   * @param {String} newValue - The new value from the autocomplete selector
+   */
+  const handleComponentTypeSelect = (e, newValue) => {
+    const selectedType = (newValue ?? e.target.value ?? "").toLowerCase();
+
+    // Generates a list of available component subtypes given a component type
+    const availableSubTypes = getAvailableSubtypes(selectedType);
+
+    // Set The selected component type
+    setSelectedComponentType(selectedType);
+    setAvailableSubtypes(availableSubTypes);
+    setSelectedComponentSubtype(null);
+  };
+
+  /**
+   * On select, it changes the value of selectedSubtype state
+   * @param {Object} e - The event object
+   * @param {String} newValue - The new value from the autocomplete selector
+   */
+  const handleComponentSubtypeSelect = (e, newValue) =>
+    setSelectedComponentSubtype(
+      (newValue ?? e.target.value ?? "").toLowerCase()
+    );
+
+  /**
+   * Retrieves the component_id based no the type and subtype names
+   * @param {string} type - The type name
+   * @param {string} subtype - The subtype name
+   */
+  const getSelectedComponentId = (type, subtype) =>
+    initialTypeCounts[type].count > 1
+      ? initialTypeCounts[type].subtypes[subtype ?? ""]?.component_id ?? null
+      : initialTypeCounts[type]?.component_id ?? null;
```

Comment:
based on the question above, if the count is >1, wont the component_id always be 0 ?

Reply:
@chiaberry almost always, except whenever we have single types without subtypes.